Updating the Data Sources docs (#6847)
apollographql/apollo-server: docs/source/data/fetching-rest.mdx

---

> title: Fetching from REST

Reviewer: Oh gosh. I forgot you had forked this file a little bit ago. @smyrick just added a bunch of stuff to the AS3 version of the data source docs that should probably be preserved! Added in https://github.com/apollographql/apollo-server/pull/6650 with a rename in https://github.com/apollographql/apollo-server/pull/6834

description: Using RESTDataSource to fetch data from REST APIs
---

The `RESTDataSource` class helps you fetch data from REST APIs. The `RESTDataSource` class helps handle caching, deduplication, and errors while resolving operations.

```mermaid
flowchart LR;
  restAPI(REST API);
  subgraph ApolloServer;
    restDataSource(RESTDataSource);
  end
  restDataSource --Fetches data--> restAPI;
  client(ApolloClient);
  client --Sends query--> ApolloServer;
```

<!-- TODO(AS4) add this note back in when fetching data article is ready
> For more information about fetching from data sources other than a REST API, see [Fetching Data](./fetching-data).
-->

> > See the [`@apollo/datasource-rest` page](https://www.npmjs.com/package/@apollo/datasource-rest) for the full details of the `RESTDataSource` API.

Reviewer: README rather than page?

## Creating subclasses

To get started, install the [`@apollo/datasource-rest` package](https://www.npmjs.com/package/@apollo/datasource-rest):

```bash
npm install @apollo/datasource-rest
```

Your server should define a separate subclass of `RESTDataSource` for each REST API it communicates with. Here's an example of a `RESTDataSource` subclass that defines two data-fetching methods, `getMovie` and `getMostViewedMovies`:
<MultiCodeBlock>

```ts title="movies-api.ts"
import { RESTDataSource } from '@apollo/datasource-rest';

class MoviesAPI extends RESTDataSource {
  override baseURL = 'https://movies-api.example.com/';

>   async getMovie(id):Promise<Movie> {

Reviewer: space after colon (same for next method) or just prettier it somehow

<MultiCodeBlock>

```ts title="index.ts"
//highlight-start
interface ContextValue {
  dataSources: {
    moviesAPI: MoviesAPI;
    personalizationAPI: PersonalizationAPI;
  };
}
//highlight-end

const server = new ApolloServer<ContextValue>({
  typeDefs,
  resolvers,
});

const { url } = await startStandaloneServer(server, {
  context: async () => {
    const { cache } = server; // highlight-line
    return {
      // We create new instances of our data sources with each request,
      // passing in our server's cache.
      //highlight-start
      dataSources: {
        moviesAPI: new MoviesAPI({ cache }),
        personalizationAPI: new PersonalizationAPI({ cache }),
      },
      //highlight-end
    };
  },
});

console.log(`🚀  Server ready at ${url}`);
```

</MultiCodeBlock>

<!-- TODO(AS4) add link to context article once exists  -->
Apollo Server calls [the `context` initialization](../resolvers/#the-context-argument) function for _every incoming operation_. This means:
- For every operation, `context` returns an _object_ containing new instances of your `RESTDataSource` subclasses (in this case, `MoviesAPI` and `PersonalizationAPI`).

> - The **`context` function should create a new instance of each `RESTDataSource` subclass for each operation.** If multiple operations share a single data source instance, you might accidentally combine results from multiple operations.

Reviewer: The reason this matters is that in addition to the shared cache described in the next section, there's also a SECOND per-RESTDataSource-object cache that caches *everything* regardless of HTTP headers in-memory.  Which was completely not documented until Shane's PR!  Merging his stuff in will add a bit more of that background but it would be nice if we actually described the two-level cache in the docs explicitly somewhere... see https://github.com/apollographql/apollo-server/pull/6834#pullrequestreview-1079413629

Author: Does it feel okay to push back documenting this to the bigger article refresh, so we can merge this faster? If not, I can get on this!

Reviewer: I mean it's up to you. I feel like this sentence raises more questions than it answers though... Maybe don't even attempt to explain why you need a new instance for each operation if you're not going to provide the background for anybody to understand?

I am not sure if this is a serious suggestion or not:

> - The **`context` function should create a new instance of each `RESTDataSource` subclass for each operation.** The reason for this is related to the caching data model used by `RESTDataSource`, which is not currently documented but we intend to in the future.

Author: Going down that road, what if we just use the first sentence for now? Omit the explanation for the moment until I can expand on this further in a (hopefully fast) follow-up:

> - The **`context` function should create a new instance of each `RESTDataSource` subclass for each operation.**

Reviewer: Yeah I think so.

Your resolvers can then access your data sources from the shared `context` object and use them to fetch data:

```ts title="resolvers.ts"
const resolvers = {
  Query: {
    movie: async (_, { id }, { dataSources }) => {
      return dataSources.moviesAPI.getMovie(id);
    },
    mostViewedMovies: async (_, __, { dataSources }) => {
      return dataSources.moviesAPI.getMostViewedMovies();
    },
    favorites: async (_, __, { dataSources }) => {
      return dataSources.personalizationAPI.getFavorites();
    },
  },
};
```

> ## Caching

Reviewer: it is probably important to mention that this cache respects headers like cache-control from the API and doesn't cache things that the backend doesn't say to cache!

Author: Great point; I added a sentence above telling folks to enable caching:

> If you've set up [caching](../preformance/caching/#recommended-starting-usage) for your server, a data source can help store the results of past fetches for particular fields in your schema.

As well as added a note after the first code snippet:

> Note that `RESTDataSource` only caches data for the fields specified by your server. To learn more about how to set up caching for your server, see [Server-side caching](../preformance/caching).

Reviewer: That's not really it. It's that it will cache data that the *REST API* specifies as cacheable in its HTTP response headers. This has nothing to do with fields or resolvers or schema or directives, because RESTDataSource is a caching HTTP client that has literally nothing to do with GraphQL or Apollo Server except for the fact that it is one way of solving a problem that many Apollo Server users do have.

Author: Okay, I obviously did not understand how that worked; thank you for explaining that! 

How does this sentence feel instead:

> The `RESTDataSource` class can cache results if the REST API it fetches from specifies caching headers in its HTTP responses (e.g., [`cache-control`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control)).

Reviewer: You probably didn't understand it because we never bothered to describe what the entire point of this project is anywhere!

That sentence looks great.

    return {
      //highlight-start
      dataSources: {
        personalizationApi: new PersonalizationAPI({ cache, token }),
      },
      //highlight-end
    };
  },
});
```

</MultiCodeBlock>

When running multiple instances of your server, you should use a shared cache backend. This enables one server instance to use the cached result from _another_ instance.

> If you want to configure or replace Apollo Sever's default cache, see [Configuring external caching](../performance/cache-backends) for more details.

## HTTP Methods

`RESTDataSource` includes convenience methods for common REST API request methods: `get`, `post`, `put`, `patch`, and `delete` ([see the source](https://github.com/apollographql/datasource-rest/blob/25862e18d8b35e324c150654c5686ed317b3fca8/src/RESTDataSource.ts#L163)).

An example of each is shown below:

<ExpansionPanel title="Click to expand">

```ts
class MoviesAPI extends RESTDataSource {
  override baseURL = 'https://movies-api.example.com/';

  // GET
  async getMovie(id) {
    return this.get(
      `movies/${encodeURIComponent(id)}`, // path
    );
  }

  // POST
  async postMovie(movie) {
    return this.post(
      `movies`, // path

>       movie, // request body

Reviewer: ```suggestion
      { body: { movie } }, // request body
```

Oops, this one too.

    );
  }

  // DELETE
  async deleteMovie(movie) {
    return this.delete(
      `movies/${encodeURIComponent(movie.id)}`, // path
    );
  }
}

```

</ExpansionPanel>

Note the use of [`encodeURIComponent`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent) in the above snippet. This is a standard function that encodes special characters in a URI, preventing a possible injection attack vector.

For a simple example, suppose our REST endpoint responded to the following URLs:

- DELETE `/movies/:id`
- DELETE `/movies/:id/characters`

A "malicious" client could provide an `:id` of `1/characters` to target the delete `characters` endpoint when it was the singular `movie` endpoint that we were trying to delete. URI encoding prevents this kind of injection by transforming the `/` into `%2F`. This can then be correctly decoded and interpreted by the server and won't be treated as a path segment.

### Method parameters

For all HTTP convenience methods, the **first parameter** is the relative path of the endpoint you're sending the request to (e.g., `movies`).

The **second parameter** depends on the HTTP method:

> - For HTTP methods with a request body (`post`, `put`, `patch`), the second parameter is an _object_ that contains the `method` `cacheOptions`, and request `body` (e.g., `{ body: { id: movie.id } }`).

Reviewer: it actually can contain a lot of stuff like `headers` too?  also missing comma after `method`

Author: Okay, you were so right—my bad for not rechecking the args! I replaced this whole section with:

> ### Method parameters
> For all HTTP convenience methods, the first parameter is the relative path of the endpoint you're sending the request to (e.g., `movies`). The second parameter is an object where you can set a request's `headers`, `params`, and `cacheOptions`, and `body` (e.g., `{ body: { id: movie.id } }`):

```ts
class MoviesAPI extends RESTDataSource {
  override baseURL = 'https://movies-api.example.com/';
  
  // an example making an HTTP POST request
  async postMovie(movie) {
    return this.post(
      `movies`, // path
      { body: movie }, // request body
    );
  }
}
```

> - For HTTP methods _without_ a request body, the second parameter is an object with keys and values corresponding to the request's query parameters.

Reviewer: Is that actually true? I think it's a similar object which can have `headers` and such, and params go on `params`?

Author: (this is also addressed above)

## Intercepting fetches

> > **New in Apollo Server 4**: Apollo Server 4 now uses [`@apollo/fetcher`](../migration#apolloutilsfetcher-replaces-apollo-server-env) under-the-hood for fetching, so hooks like `willSendRequest` no longer have direct access to `Request` or `Headers` methods.

Reviewer: ```suggestion
> **New in Apollo Server 4**: Apollo Server 4 now uses the [`@apollo/utils.fetcher`](../migration#apolloutilsfetcher-replaces-apollo-server-env) interface under the hood for fetching. This interface lets you choose your own implementation of the Fetch API. To ensure compatibility with all Fetch implementations, the request provided to hooks like `willSendRequest` is a plain JS object rather than a `Request` object with methods.
```

maybe something like this?

  constructor() {
    super();
  }

  async getMovie(id):Promise<Movie> {
    return this.get(`movies/${encodeURIComponent(id)}`);
  }

  async getMostViewedMovies(limit = '10'):Promise<Movie[]> {
    const data = await this.get('movies', {
      params: {
        per_page: limit,
        order_by: 'most_viewed',
      },
    });
    return data.results;
  }
}
```

</MultiCodeBlock>

You can extend the `RESTDataSource` class to implement whatever data-fetching methods your resolvers need. These methods can use built-in convenience methods (like `get` and `post`) to perform HTTP requests, helping you add query parameters, parse JSON results, and handle errors.

## Adding subclasses to Apollo Server's context

You can then provide your `RESTDataSource` subclasses to the `context` initialization function, like so:

<MultiCodeBlock>

```ts title="index.ts"
//highlight-start
interface ContextValue {
  dataSources: {
    moviesAPI: MoviesAPI;
    personalizationAPI: PersonalizationAPI;
  };
}
//highlight-end

const server = new ApolloServer<ContextValue>({
  typeDefs,
  resolvers,
});

const { url } = await startStandaloneServer(server, {
  context: async ({ req }) => {

>     return {

Reviewer 2: We want to pass the Apollo Server cache to the data sources when we instantiate them. Assuming we haven't overridden the constructor in our subclass implementations, they'll share the same signature as `RESTDataSource` itself.

```ts
const { url } = await startStandaloneServer(server, {
  context: async ({ req }) => {
    const { cache } = server;
    return {
      //highlight-start
      // We create new instances of our data sources with each request
      dataSources: {
        moviesAPI: new MoviesAPI({ cache }),
        personalizationAPI: new PersonalizationAPI({ cache }),
      },
      //highlight-end
    };
  },
});
```

      //highlight-start
      // We create new instances of our data sources with each request
      dataSources: {
        moviesAPI: new MoviesAPI(),
        personalizationAPI: new PersonalizationAPI(),
      },
      //highlight-end
    };
  },
});

console.log(`🚀  Server ready at ${url}`);
```

</MultiCodeBlock>

<!-- TODO(AS4) add link to context article once exists  -->
Apollo Server calls the [the `context` initialization](../resolvers/#the-context-argument) function for _every incoming operation_. This means:
- As shown above, with every operation `context` returns an _object_ containing new instances of your `RESTDataSource` subclasses (in this case, `MoviesAPI` and `PersonalizationAPI`).

> -  These subclasses are accessible from the [the `context` argument](./resolvers/#the-context-argument) that's passed between your server's resolvers.

Reviewer 2: Are we calling this `contextValue` everywhere or no? We might need to rename the heading, on the resolvers doc, etc.

Author: Yeah, I'm going to do that in a separate PR after merging in all the AS4 docs in, so I can change `contextValue` everywhere all at once. I think it'll be less error-prone that way  😅

## Caching

> 📣 **New in Apollo Server 4**: Apollo Server no longer automatically provides it's cache to data sources. [See here for more details](#datasources).

> You can pass your `RESTDataSource` subclasses a cache (such as Apollo Server's default cache) to store the results of past fetches:

Reviewer 2: 👍

<MultiCodeBlock>

```ts
import { RESTDataSource } from '@apollo/datasource-rest';
// KeyValueCache is the type of Apollo server's default cache

> import { KeyValueCache } from '@apollo/utils.keyvaluecache';

Reviewer 2: We _can_ use type imports for this (though I don't know that we do that consistently in our TS docs). This is totally optional.
```suggestion
import type { KeyValueCache } from '@apollo/utils.keyvaluecache';
```

Author: Oh, rad, I'll try to do that from now on!

import { ApolloServer } from '@apollo/server';
import { startStandaloneServer } from '@apollo/server/standalone';

class MoviesAPI extends RESTDataSource {
  baseURL = 'https://movies-api.example.com/';

  //highlight-start
  constructor(options: {cache: KeyValueCache }) { 
    super(options); // this should send our server's `cache` through
  }

>   //highlight-end

Reviewer 2: This whole constructor can be omitted (it's the default behavior if we don't implement the constructor).

However, this also might be interesting for illustrating what's going on and what's available to your constructor. If we do leave this here, it'd be worth mentioning this happens automatically without it there.

Reviewer 2: See comment below in conjunction with this comment

  // data fetching methods, etc.
}

interface ContextValue {
  dataSources: {
    moviesAPI: MoviesAPI;
  };
}

const server = new ApolloServer<ContextValue>({
  typeDefs,
  resolvers,
});

const { url } = await startStandaloneServer(server, {
  context: async ({ req }) => {
    const { cache } = server;
    return {
      token,
      //highlight-start
      dataSources: {

>         moviesAPI: new MoviesAPI({ cache, token }),

Reviewer 2: If this is referencing the `MoviesAPI` we created above, then we need to include the token in the constructor signature.

And if we do that, then my previous comment about dropping the default constructor is no longer valid (since `token` is not a part of the base class's constructor signature).

      },
      //highlight-end
    };
  },
});
```

</MultiCodeBlock>

> If you want Apollo Server to use another cache backend, you can provide its constructor a _different_ cache object that implements the [`KeyValueCache` interface](https://github.com/apollographql/apollo-utils/tree/main/packages/keyValueCache#keyvaluecache-interface):

Reviewer 2: Custom cache for Apollo Server / external caches should definitely go on a different page from RESTDataSource stuff.

Author: Great point; I added a callout with a link to the Configuring Cache backends article. :)

```ts title="server.ts"
const server = new ApolloServer({
  typeDefs,
  resolvers,
  cache: new MyCustomKeyValueCache()
});
```

### Using an external cache backend

When running multiple instances of your server, you should use a shared cache backend. This enables one server instance to use the cached result from _another_ instance.

Apollo Server supports using [Memcached](https://memcached.org/), [Redis](https://redis.io/), or other cache backends via the [`keyv`](https://www.npmjs.com/package/keyv) package. For examples, see [Configuring external caching](../performance/cache-backends#configuring-external-caching).

You can also choose to implement your own cache backend. For more information, see [Implementing your own cache backend](../performance/cache-backends#implementing-your-own-cache-backend).

## HTTP Methods

> `RESTDataSource` includes convenience methods for common REST API request methods: `get`, `post`, `put`, `patch`, and `delete` ([see the source](https://github.com/apollographql/datasource-rest/blob/main/src/RESTDataSource.ts#L163)).

Reviewer 2: If we're linking to a specific line number, we _shouldn't_ reference `.../blob/main/...` since that's a moving target. If the code changes on main, your link becomes invalid.

To remedy this, when you select the line in GH you'll see 3 dots - you want copy permalink. You can also just replace `main` in your URL with the most recent commit hash.

https://github.com/apollographql/datasource-rest/blob/25862e18d8b35e324c150654c5686ed317b3fca8/src/RESTDataSource.ts#L163

Author: That's a great point, I always do wonder what the best way to do GH code links is. I'd rather not link folks to old commits, just in case you update things, but I can't think of a better way right now—so permalink it is!

An example of each is shown below:

<ExpansionPanel title="Click to expand">

```ts
class MoviesAPI extends RESTDataSource {

>   baseURL = 'https://movies-api.example.com/';

Reviewer 2: ```suggestion
  override baseURL = 'https://movies-api.example.com/';
```

A "malicious" client could provide an `:id` of `1/characters` to target the delete `characters` endpoint when it was the singular `movie` endpoint that we were trying to delete. URI encoding prevents this kind of injection by transforming the `/` into `%2F`. This can then be correctly decoded and interpreted by the server and won't be treated as a path segment.

### Method parameters

For all HTTP convenience methods, the **first parameter** is the relative path of the endpoint you're sending the request to (e.g., `movies`).

The **second parameter** depends on the HTTP method:

> - For HTTP methods with a request body (`post`, `put`, `patch`), the second parameter _is_ the request body (e.g., `{ body: { id: movie.id } }`).

Reviewer 2: The second parameter is actually an object which contains the request body (which you've correctly illustrated in your example). It also includes `method` and potentially `cacheOptions`.

Author: Oh, thank you, I forgot to correct the sentence before the example 🤦‍♀️

## Intercepting fetches

`RESTDataSource` includes a `willSendRequest` method that you can override to modify outgoing requests before they're sent. For example, you can use this method to add headers or query parameters. This method is most commonly used for authorization or other concerns that apply to all sent requests.

Data sources also have access to the GraphQL operation context, which is useful for storing a user token or other relevant information.

### Setting a header

<MultiCodeBlock>

```ts
import { RESTDataSource, WillSendRequestOptions } from '@apollo/datasource-rest';

class PersonalizationAPI extends RESTDataSource {
  override baseURL = 'https://personalization-api.example.com/';

  constructor(private token: string) {
    super();
  }

  override willSendRequest(request: WillSendRequestOptions) {
    request.headers['authorization'] = this.token;
  }
}
```

</MultiCodeBlock>

> > If you're using TypeScript, make sure to import the `WillSendRequestOptions` type.

Reviewer 2: Should we just do this in our example?

Author: Yes, yes we should!

- Can be shared across multiple GraphQL servers
- Provides cache management features like expiry and invalidation that use standard HTTP cache control headers

### Batching with REST APIs

Most REST APIs don't support batching. When they do, using a batched endpoint can _jeopardize_ caching. When you fetch data in a batch request, the response you receive is for the exact combination of resources you're requesting. Unless you request that same combination again, future requests for the same resource won't be served from cache.

We recommend that you restrict batching to requests that _can't_ be cached. In these cases, you can take advantage of DataLoader as a private implementation detail inside your `RESTDataSource`:

```ts
class PersonalizationAPI extends RESTDataSource {

>   constructor() {

Reviewer 2: We can update this to what we've been doing (remove constructor, and `override baseURL = ...`) in the class body

Reviewer 2: Pretty sure `willSendRequest` below needs an `override`
